efficientdet/keras/train_lib.py

```diff
@@ -0,0 +1,474 @@
+    ])
+
+  def _configure_callbacks(self, callbacks):
+    for callback in callbacks:
+      callback.set_model(self)
+
+  def _detection_loss(self, cls_outputs, box_outputs, labels):
+    num_positives_sum = tf.reduce_sum(labels['mean_num_positives']) + 1.0
+    levels = range(len(cls_outputs))
+    cls_losses = []
+    box_losses = []
+    box_iou_losses = []
+    for level in levels:
+      # Onehot encoding for classification labels.
+      cls_targets_at_level = tf.one_hot(labels['cls_targets_%d' % (level + 3)],
+                                        self.config.num_classes)
+
+      if self.config.data_format == 'channels_first':
+        bs, _, width, height, _ = cls_targets_at_level.get_shape().as_list()
+        cls_targets_at_level = tf.reshape(cls_targets_at_level,
+                                          [bs, -1, width, height])
+      else:
+        bs, width, height, _, _ = cls_targets_at_level.get_shape().as_list()
+        cls_targets_at_level = tf.reshape(cls_targets_at_level,
+                                          [bs, width, height, -1])
+      box_targets_at_level = labels['box_targets_%d' % (level + 3)]
+
+      class_loss = self.loss.get("class_loss", None)
+      cls_loss = class_loss([num_positives_sum, cls_outputs[level]],
+                            cls_targets_at_level)
+
+      if self.config.data_format == 'channels_first':
+        cls_loss = tf.reshape(cls_loss,
+                              [bs, -1, width, height, self.config.num_classes])
+      else:
+        cls_loss = tf.reshape(cls_loss,
+                              [bs, width, height, -1, self.config.num_classes])
+      cls_loss *= tf.cast(
+          tf.expand_dims(
+              tf.not_equal(labels['cls_targets_%d' % (level + 3)], -2), -1),
+          tf.float32)
+      cls_losses.append(tf.reduce_sum(cls_loss))
+
+      if self.config.box_loss_weight:
+        box_loss = self.loss.get("box_loss", None)
+        box_losses.append(
+            box_loss([num_positives_sum, box_outputs[level]],
+                     box_targets_at_level))
+
+      if self.config.iou_loss_type:
+        box_iou_loss = self.loss.get("box_iou_loss", None)
+        box_iou_losses.append(
+            box_iou_loss(box_outputs[level], box_targets_at_level,
+                         num_positives_sum, self.config.iou_loss_type))
+    cls_loss = tf.add_n(cls_losses)
+    box_loss = tf.add_n(box_losses) if box_losses else 0
+    box_iou_loss = tf.add_n(box_iou_losses) if box_iou_losses else 0
+    total_loss = (cls_loss + self.config.box_loss_weight * box_loss +
+                  self.config.iou_loss_weight * box_iou_loss)
+    return total_loss, cls_loss, box_loss, box_iou_loss
+
+  def _steps_loop(self, steps, dataset_iter, is_train, logs, callbacks):
+    total_loss = 0.0
+    for step in (1, steps + 1):
+      self.step.assign(step)
+      features, labels = next(dataset_iter)
+      [callback.on_batch_begin(step, logs) for callback in callbacks]
```

Comment:
list comprehension looks a little strange here because the list is not used anywhere (it will also triggler warning "expression-not-assigned".

How about change it to:

`map(lambda x, s=step: x.on_batch_begin(s, logs), callbacks)`